pinc/upload_file.inc

```diff
@@ -0,0 +1,255 @@
+            chmod($file_path, 0644);
+        }
+        else
+        {
+            // resumable upload
+            $root_staging_dir = "/tmp/resumable_uploads";
+            $original_name = array_get($_POST, "resumable_filename", "");
+            $identifier = array_get($_POST, "resumable_identifier", "");
+            $hashed_filename = md5($identifier);
+            $file_path = "$root_staging_dir/$hashed_filename";
+            $file_size = filesize($file_path);
+            $file_info = ["name" => $original_name, "tmp_name" => $file_path];
+        }
+        // do some checks.
+        empty_check($file_size);
+        virus_check($file_path);
+        return $file_info;
+    }
+    catch(FileException $e)
+    {
+        if(is_file($file_path))
+        {
+            unlink($file_path);
+        }
+        throw($e);
+    }
+}
+
+function empty_check($file_size)
+{
+    if (0 == $file_size) {
+        throw new FileException(_("File is empty"));
+    }
+}
+
+function zip_check($original_name, $file_path)
+{
+    try
+    {
+        if (substr($original_name, -4) != ".zip")
+        {
+            throw new FileException( _("File name extension must be '.zip'") );
+        }
+        // ensure that it's a valid zip
+        // The extension was already checked and the file is not properly named (it has some temporary name), so we should
+        // disable the extension check.
+        if (!is_valid_zip_file($file_path, true))
+        {
+            throw new FileException(sprintf(_("%s is not a valid zip file"),  $original_name));
+        }
+    }
+    catch(FileException $e)
+    {
+        unlink($file_path);
+        throw($e);
+    }
+}
+
+function virus_check($file_path)
+{
+    global $antivirus_executable;
+    if(!$antivirus_executable)
+    {
+        return;
+    }
+    try
+    {
+        // perform '$antivirus_executable -- <FILENAME>' and expect return
+        // value = 0. we use -- to not parse any further arguments starting
+        // with -/-- as options
+        $av_test_result = array();
+        $av_retval=0;
+
+        $cmd = "$antivirus_executable -- " . escapeshellcmd($file_path);
+        exec($cmd, $av_test_result, $av_retval);
+        // $av_retval == 0 is ok
+        if ($av_retval == 1)
+        {
+            // Log the infected upload so that we can track user/frequency
+            $reporting_string = "Infected upload: " . $av_test_result[0];
+            error_log($reporting_string);
+            throw new FileException(sprintf(_("AntiVirus: The scan reported an infection: %s. The upload has been discarded.
+                You should perform a complete virus scan on your computer as soon as possible."), $av_test_result[0]));
+        }
+        else if($av_retval > 1)
+        {
+            throw new FileException(_("AntiVirus error:") . $av_retval . "<br>" . $av_test_result[0]);
+        }
+    }
+    catch(FileException $e)
+    {
+        unlink($file_path);
```

Comment:
It feels weird to have this function remove the thing that was uploaded. I would expect the calling function to do that.